File: old/source/frontend/register_symbols.cpp
#include <config.hpp>

#include <iostream>

// Find module
static Input *find_module(String name)
{
    for (Input &input : g_inputs)
    {
        if (input.module_name == name)
            return &input;
    }

    return nullptr;
}

// Register symbols
void Node::register_symbols(Input *input)
{

}

// Register symbols
void NodeImport::register_symbols(Input *input)
{
    // Find module
    Input *mod = find_module(name);

    if (!mod)
        input->logger.error(start, "This module does not exists.");
    
    // Add module symbols
    mod->register_symbols();

    for (auto &pair : mod->globals)
    {
        if (input->globals.find(pair.first) != input->globals.end())
            mod->logger.error("SYMBOL DECLARED.....");

        input->globals[pair.first] = pair.second;
    }
}

// Register symbol
void NodeFunction::register_symbols(Input *input)
{
    // Check for function redeclaration
    if (input->globals.find(name.to_string()) != input->globals.end())
        input->logger.error(name, "This function was already declared.");

    // Add function symbol
    input->globals[name.to_string()] = { SymbolKind::Function };
}

File: source/frontend/passes/RegisterSymbols.cs
using System.Collections.Generic;

namespace Suteki
{
    partial class NodeImport : Node
    {
        public override void RegisterSymbols(Input input)
        {
            // Find module
            Module module = Config.Modules[ModuleName.AsString()];

            if (module == null)
                input.Logger.Error(Start, "This module was not found.");
            else
                input.Imports.Add(module);
        }
    }

    partial class NodeFunction : Node
    {
        public override void RegisterSymbols(Input input)
        {
            // Check for existing symbol
            if (input.Find(Name))
            {
                input.Logger.Error(Name, "This symbol was already declared.");
                return;
            }

            // Add symbol
            input.Module.Symbols.Add(Name.Content, new Symbol(Property, SymbolKind.Function, this));
        }
    }
}